Implementation of new Batching API

rahulKQL opened the pull request.

rahulKQL committed:
Introducing new Batching API for Bigtable

 - Introducing new API `BigtableDataClient#newBulkMutationBatcher(String)`.
 - Deprecated existing API `BigtableDataClient#newBulkMutationBatcher()`.
 - Updated BatchingCallSettings to be compatible with new Batching API.
 - Removed method/class description from deprecation.
 - Added a new model in RowMutationEntry.
 - Added MutateRowsBatchingDescriptorV2 to spool the batching response.
 - Updated JavaDocs, test cases according to latest changes.

igorbernstein2 commented on google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/BigtableDataClient.java:
```diff
@@ -886,7 +889,10 @@ public void mutateRow(RowMutation rowMutation) {
    * }
    * }</pre>
+   *
+   * @deprecated Please use latest {@link #newBulkMutationBatcher(String)} API.
```
Please remove the other docs and only retain the deprecation notice

rahulKQL replied:
I have kept the method description Shall we remove those as well?

igorbernstein2 replied:
yes please

igorbernstein2 commented on google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/models/BulkMutation.java:
```diff
@@ -100,4 +100,10 @@ public MutateRowsRequest toProto(RequestContext requestContext) {
         .build();
   }
+
+  public BulkMutation clone() {
```
`@Override`

rahulKQL replied:
Thanks for pointing it out.

igorbernstein2 commented on google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/models/BulkMutationBatcher.java:
```diff
@@ -33,7 +33,11 @@
  * @see com.google.cloud.bigtable.data.v2.BigtableDataClient#newBulkMutationBatcher() for example
  *     usage.
+ *
+ * @deprecated @deprecated Please use {@link
```
- please remove the 2nd deprecated
- also please remove all other docs other then deprecation notice

igorbernstein2 commented on google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/models/RowMutationEntry.java:
```diff
@@ -0,0 +1,133 @@
+  }
+
+  @Override
+  public RowMutationEntry setCell(
```
Please inherit javadoc

igorbernstein2 commented on google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/EnhancedBigtableStubSettings.java:
```diff
@@ -389,7 +389,8 @@ public Builder toBuilder() {
     private final UnaryCallSettings.Builder<String, List<KeyOffset>> sampleRowKeysSettings;
     private final UnaryCallSettings.Builder<RowMutation, Void> mutateRowSettings;
-    private final BatchingCallSettings.Builder<RowMutation, Void> bulkMutateRowsSettings;
+    private final BatchingCallSettings.Builder<RowMutationEntry, Void, RowMutation, Void>
```
Why is the request type RowMutation? Shouldn't it be BulkMutation?

rahulKQL replied:
Actually, I mistakenly assumed that we should not deprecate and change the Request wrapper RowMutation to BulkMutation in one go.

But I have updated it after you pointed it out.

igorbernstein2 commented on google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/EnhancedBigtableStubSettings.java:
```diff
@@ -501,36 +502,26 @@ private Builder(EnhancedBigtableStubSettings settings) {
      * constructing the callable chain.
      */
     private static class PlaceholderBatchingDescriptor
```
I don't think the placeholder should be necessary anymore

rahulKQL replied:
Removed the same.

kolea2 commented on google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/EnhancedBigtableStub.java:
```diff
@@ -75,6 +79,8 @@
 public class EnhancedBigtableStub implements AutoCloseable {
   private static final String TRACING_OUTER_CLIENT_NAME = "Bigtable";
   private static final String TRACING_INNER_CLIENT_NAME = "BaseBigtable";
+  private static final MutateRowsBatchingDescriptorV2 batchingDescriptorV2 =
```
is this going to live side by side with MutateRowsBatchingDescriptor? Can we replace the old one with this?

rahulKQL replied:
Thanks for the review.
Yes, We are aiming to remove existing `MutateRowsBatchingDescriptor`. I have added a TODO in `MutateRowsBatchingDescriptorV2` for the same reason.

kolea2 commented on google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/mutaterows/MutateRowsBatchingDescriptorV2.java:
```diff
@@ -0,0 +1,125 @@
+ * applications directly.
+ */
+// TODO(rahulkql): Consider renaming this class
+public class MutateRowsBatchingDescriptorV2
```
can we remove "public" if it is internal only? If no, let's add `@InternalApi`

rahulKQL replied:
We would need this class in other packages so I have marked it as `@InternalApi("For internal use only")`.

rahulKQL commented:
@igorbernstein2 @kolea2 
My apologies for the late response, I have addressed the feedback comments. 

Could you please have a fresh look at this PR whenever have time.

igorbernstein2 commented on google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/BigtableDataClient.java:
```diff
@@ -921,6 +909,34 @@ public void bulkMutateRows(BulkMutation mutation) {
+   * <pre>{@code
+   * try (BigtableDataClient bigtableDataClient = BigtableDataClient.create("[PROJECT]", "[INSTANCE]")) {
+   *   try (Batcher<RowMutationEntry, Void> batcher = bigtableDataClient.newBulkMutationBatcher("[TABLE]")) {
+   *     for (String someVale : someCollections) {
```
`someValue : someCollection`

igorbernstein2 commented on google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/models/BulkMutation.java:
```diff
@@ -34,7 +34,7 @@
  * automatic batching with flow control.
  */
-public final class BulkMutation implements Serializable {
+public final class BulkMutation implements Serializable, Cloneable {
```
Please add a test to make sure that clone doesn't share data with the original and has all of the necessary attributes

rahulKQL replied:
I have added a test case for this in `BulkMutation#cloneTests()`.

igorbernstein2 commented on google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/models/RowMutationEntry.java:
```diff
@@ -0,0 +1,142 @@
+import javax.annotation.Nonnull;
+
+/**
+ * Represents a list of mutations targeted at a single row. It's meant to be used as an parameter
```
Please add a comment that the mutations will be applied atomically and in order

igorbernstein2 commented on google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/models/RowMutationEntry.java:
```diff
@@ -0,0 +1,142 @@
+  }
+
+  /** Creates a new instance of the mutation builder. */
+  public static RowMutationEntry create(@Nonnull String key, @Nonnull Mutation mutation) {
```
Is it necessary to expose this?

rahulKQL replied:
I don't think it is necessary, I copied these to have consistency between `RowMutation`

igorbernstein2 replied:
I think those were added to RowMutation for explicit usage by the hbase client

rahulKQL replied:
Thanks for clarifying it, I have removed those from `RowMutationEntry`.

igorbernstein2 commented on google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/models/RowMutationEntry.java:
```diff
@@ -0,0 +1,142 @@
+  }
+
+  /** Creates a new instance of the mutation builder. */
+  public static RowMutationEntry create(@Nonnull ByteString key, @Nonnull Mutation mutation) {
```
Is it necessary to expose this?

igorbernstein2 commented on google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/EnhancedBigtableStub.java:
```diff
@@ -477,7 +495,10 @@ public static EnhancedBigtableStub create(EnhancedBigtableStubSettings settings)
   /**
    * Returns the callable chain created in {@link #createBulkMutateRowsBatchingCallable()} ()}
    * during stub construction.
+   *
```
Please remove doc except for the deprecation notice.

igorbernstein2 commented on google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/mutaterows/MutateRowsBatchingDescriptorV2.java:
```diff
@@ -0,0 +1,127 @@
+    return request.toProto().getSerializedSize();
+  }
+
+  /** A {@link BatchingRequestBuilder} that can aggregate BulkMutation */
```
that will spool mutations and send them out as a `{@link BulkMutation}`

igorbernstein2 commented on google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/mutaterows/MutateRowsBatchingDescriptorV2.java:
```diff
@@ -0,0 +1,127 @@
+      // NOTE: The gax batching api doesn't allow for a single response to get different
+      // exception for different entries. However this does not affect this client because
+      // SettableApiFuture corresponds to  single element. So just use the last error per entry.
+      ApiException lastError = null;
```
I dont understand what you mean here. How can a single entry get multiple errors?

rahulKQL replied:
I was trying to map FailedMutation to each entry, I had assumed that `MutateRowsException#getFailedMutation()` could return different number of FailedMutation than the number of entries on that batch.

In case the number of entries exceeds the number of FailedMutation, we are setting the last failedMutation to the rest of the entries.

Should we set throwable to each entry (without checking MutateRowsException)? Something like this:
```java
  ...
  @Override
  public void splitException(Throwable throwable, List<SettableApiFuture<Void>> batch) {
    for (SettableApiFuture<Void> future : batch) {
      future.setException(throwable);
    }
  }
  ...
```

igorbernstein2 replied:
The first approach is correct (unpacking the entry errors from `MutateRowsException#getFailedMutations`). However the comment is misleading. Each entry will have at most 1 error. In fact if there were duplicate keys then `Maps.uniqueIndex` would throw an IllegalArgumentException. To be honest I think this would be easier to read:

```java
Map<Integer, Throwable> entryErrors = new HashMap<>();
for(FailedMutation failure : failedMutations) {
  entryErrors.put(failure.getIndex(), failure.getError();
}

int i = 0;
for (SettableApiFuture<Void> entryResultFuture : batch) {
  Throwable entryError = entryErrors.get(i++);
  if (entryError == null) {
     entryResultFuture.set(null);
  } else {
     entryResultFuture.setException(entryError);
  }
}    
```

rahulKQL replied:
Thanks a lot for simplifying it. I have updated this method with the same approach.

igorbernstein2 commented on google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/mutaterows/MutateRowsBatchingDescriptorV2.java:
```diff
@@ -0,0 +1,127 @@
+    public void add(RowMutationEntry rowEntry) {
+      MutateRowsRequest.Entry entry = rowEntry.toProto();
+      ByteString rowK = entry.getRowKey();
+      bulkMutation.add(rowK, Mutation.fromProtoUnsafe(entry.getMutationsList()));
```
Can BulkMutation take an instance of RowMutationEntry?

rahulKQL replied:
Added a `BulkMutation#add(RowMutationEntry)`, I hope that should be fine.

igorbernstein2 commented on google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/mutaterows/MutateRowsBatchingDescriptorV2.java:
```diff
@@ -0,0 +1,127 @@
+  }
+
+  /** {@inheritDoc} */
+  public long countBytes(RowMutationEntry request) {
```
s/request/entry

igorbernstein2 commented on google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/mutaterows/MutateRowsBatchingDescriptorV2.java:
```diff
@@ -0,0 +1,127 @@
+    }
+
+    @Override
+    public void add(RowMutationEntry rowEntry) {
```
Please be consistent in your variable names

rahulKQL replied:
Thanks, Will now onwards try to be consistent.

igorbernstein2 commented on google-cloud-clients/google-cloud-bigtable/src/test/java/com/google/cloud/bigtable/data/v2/it/RowMutationEntryBatcherIT.java:
```diff
@@ -0,0 +1,72 @@
+    }
+    ServerStream<Row> actualRows = client.readRows(Query.create(tableId).prefix(rowPrefix));
+
+    Truth.assertThat(actualRows).containsExactlyElementsIn(expectedRows);
```
static import please

igorbernstein2 commented on google-cloud-clients/google-cloud-bigtable/src/test/java/com/google/cloud/bigtable/data/v2/stub/EnhancedBigtableStubSettingsTest.java:
```diff
@@ -418,7 +420,7 @@ public void bulkMutateRowsSettingsAreNotLostTest() {
         .containsAtLeast(Code.ABORTED, Code.DEADLINE_EXCEEDED);
     assertThat(builder.bulkMutateRowsSettings().getRetrySettings()).isEqualTo(retrySettings);
-    assertThat(builder.bulkMutateRowsSettings().getBatchingSettings())
+    assertThat(builder.bulkMutateRowsSettings().build().getBatchingSettings())
```
why is this change necessary?

rahulKQL replied:
We don't have getters anymore in new BatchingCallSettings.Builder. So I had an option either to review this assertion or build and get the BatchingSettings.

igorbernstein2 replied:
I would just add a getter to BatchingCallSettings

rahulKQL replied:
Did you mean add a getter to `BatchingCallSettings.Builder`?

igorbernstein2 replied:
yep

kolea2 commented on google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/mutaterows/MutateRowsBatchingDescriptorV2.java:
```diff
@@ -0,0 +1,114 @@
+   * A {@link BatchingRequestBuilder} that will spool mutations and send them out as a {@link
+   * BulkMutation}.
+   */
+  static class MyRequestBuilder implements BatchingRequestBuilder<RowMutationEntry, BulkMutation> {
```
minor - let's rename this. RequestBuilder maybe?

rahulKQL committed:
Addressing minor feedback comments

 - Fix for linkage error on Javadocs
 - Introduced BigtableBatchingCallSettings which will wrap the underlying gax's BatchingCallSettings. This was done for enabling the bigtable throttling to balance the bigtable clusters performance supposedly for batch jobs while still supporting the live load.
 - Added null check message for RowMutationEntry
 - Updated javadoc and access level of getter BigtableBatchingCallSettings

rahulKQL committed:
fix for Linkage monitor

This is a trial fix, as linkage monitor does not seems to be recreated with local build

rahulKQL committed:
Addressing feedback comments mostly for javadoc and variable names

rahulKQL committed:
Annotate BatchingCallSettings with `@Nonnull`

rahulKQL commented:
Closing this PR as this is already merged in origin repo.
